### PR TITLE
GC+/ Milestone 2 execution plan

### LINKED ISSUE #4802: GC+/Milestone 2 execution plan

### PR DESCRIPTION
**Pull request by eden-ohana**
closes #4802

### CODE REVIEW

File: design/accepted/gc_plus/gc-plus-milestone2-execution-plan.md [hunk 1]
```diff
@@ -0,0 +1,50 @@
+# Uncommitted Garbage Collection - Milestone 2 Execution Plan
+
+Uncommitted Garbage Collection [Proposal](https://github.com/treeverse/lakeFS/blob/master/design/accepted/gc_plus/uncommitted-gc.md)
+
+## Milestone 1
+The first beta version that was released included:
```
Comment (nopcoder) [inline design/accepted/gc_plus/gc-plus-milestone2-execution-plan.md:6]:
```suggestion
## Milestone 1

The first beta version that was released included:
```

File: design/accepted/gc_plus/gc-plus-milestone2-execution-plan.md [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+1. Implementation of the clean run flow for old and new repository structures (without optimizations)
+2. Mark & Sweep
+3. Integration tests
+4. Backup & Restore
```
Comment (nopcoder) [inline design/accepted/gc_plus/gc-plus-milestone2-execution-plan.md:10]: minimal support using rclone

File: design/accepted/gc_plus/gc-plus-milestone2-execution-plan.md [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+
+## Milestone 2
+
+### Goals
+1. Removing the limitation of a read-only lakeFS during the job run
```
Comment (nopcoder) [inline design/accepted/gc_plus/gc-plus-milestone2-execution-plan.md:15]: job run == gc?

File: design/accepted/gc_plus/gc-plus-milestone2-execution-plan.md [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+2. Performance improvements - better parallelization of the storage namespace traversal
+3. Implementing Optimized Run
```
Comment (nopcoder) [inline design/accepted/gc_plus/gc-plus-milestone2-execution-plan.md:17]: 2 and 3 are performance and optimizations - the first item is self explained, I didn't understand the second one.
Do you mean incremental uncommitted gc?

File: design/accepted/gc_plus/gc-plus-milestone2-execution-plan.md [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+1. Required changes by lakeFS:
+    * [[Get/Link]PhysicalAddress](https://github.com/treeverse/lakeFS/issues/4476)
+        * [Validation of cutoff time](https://github.com/treeverse/lakeFS/issues/4695)
+    * [StageObject API](https://github.com/treeverse/lakeFS/issues/4480)
+    * [CopyObject API](https://github.com/treeverse/lakeFS/issues/4477)
+        * [S3 Gateway CopyObject](https://github.com/treeverse/lakeFS/issues/4478)
+        * [lakeFSFS renameObject method](https://github.com/treeverse/lakeFS/issues/4479)
+    * [Track copied objects in ref-store](https://github.com/treeverse/lakeFS/issues/4562)
+
+3. Performance improvements:
```
Comment (nopcoder) [inline design/accepted/gc_plus/gc-plus-milestone2-execution-plan.md:39]: These are all uncommitted GC? maybe we can specify it as there are lakefs, common code and etc.
```suggestion
2. Performance improvements:
```

File: design/accepted/gc_plus/gc-plus-milestone2-execution-plan.md [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+    * [Optimaized listing on old repository structure](https://github.com/treeverse/lakeFS/issues/4620)
+    * [Efficient listing on committed entries](https://github.com/treeverse/lakeFS/issues/4600)
+    * [Optimized listing on slices](https://github.com/treeverse/lakeFS/issues/4614)
```
Comment (nopcoder) [inline design/accepted/gc_plus/gc-plus-milestone2-execution-plan.md:42]: Thought we already doing this one

File: design/accepted/gc_plus/gc-plus-milestone2-execution-plan.md (hunk 4 above)
Comment (N-o-Z) [reply design/accepted/gc_plus/gc-plus-milestone2-execution-plan.md:17]: @nopcoder Optimized run means incremental
@eden-ohana I think we should focus this milestone on optimization of the full run and adding more testing (unit / integration) - and create a separate milestone for the incremental/optimized run

File: design/accepted/gc_plus/gc-plus-milestone2-execution-plan.md [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+
+## Milestone 2
+
+### Goals
+1. Removing the limitation of a read-only lakeFS during the job run
+2. Performance improvements - better parallelization of the storage namespace traversal
+3. Implementing Optimized Run
+
+### Non-Goals
+1. Support for non-S3 repositories
```
Comment (N-o-Z) [inline design/accepted/gc_plus/gc-plus-milestone2-execution-plan.md:20]: This is a big bullet - I think it should be itemized by storage providers. Also I really think we should do it incremental.
For example - for this milestone, starting with Azure - and afterwards working on GCS

File: design/accepted/gc_plus/gc-plus-milestone2-execution-plan.md (hunk 6 above)
Comment (N-o-Z) [reply design/accepted/gc_plus/gc-plus-milestone2-execution-plan.md:42]: @nopcoder  Yes you did :) I think we can close this issue

**Review by N-o-Z — CHANGES_REQUESTED** [review]
@eden-ohana Thanks for creating this plan. 
It looks well put through - I think we learned a lot from the first milestone, especially about our expectations and we should put this into this plan.
There is a lot of work hidden inside the milestone goals lines in my  opinion.
I believe we should focus on improving our current solution for the 2nd milestone - structuring it, creating more tests and add optimizations.
We should also start to create the ground for incremental runs but I think the client implementation itself should be planned for milestone 3

File: design/accepted/gc_plus/gc-plus-milestone2-execution-plan.md (hunk 1 above)
Comment (eden-ohana) [reply design/accepted/gc_plus/gc-plus-milestone2-execution-plan.md]: done

File: design/accepted/gc_plus/gc-plus-milestone2-execution-plan.md (hunk 2 above)
Comment (eden-ohana) [reply design/accepted/gc_plus/gc-plus-milestone2-execution-plan.md]: done

File: design/accepted/gc_plus/gc-plus-milestone2-execution-plan.md (hunk 3 above)
Comment (eden-ohana) [reply design/accepted/gc_plus/gc-plus-milestone2-execution-plan.md]: yes

File: design/accepted/gc_plus/gc-plus-milestone2-execution-plan.md (hunk 5 above)
Comment (eden-ohana) [reply design/accepted/gc_plus/gc-plus-milestone2-execution-plan.md]: done

File: design/accepted/gc_plus/gc-plus-milestone2-execution-plan.md (hunk 6 above)
Comment (eden-ohana) [reply design/accepted/gc_plus/gc-plus-milestone2-execution-plan.md]: removed

**Comment by eden-ohana** [comment]
@ozkatz can you give context from a product perspective? Is there any requirement to prioritize the GC+ for Azure?

**Comment by ozkatz** [comment]
@eden-ohana I was under the impression that since the "Sweep" part of GC+ is re-used (i.e. it's the same "sweep" that is already implemented for Azure) we should support this ootb?

There are active, production installations that we _know_ of that are running on Azure, with compliance requirements that are very hard to meet without GC+..

**Comment by eden-ohana** [comment]
> @eden-ohana I was under the impression that since the "Sweep" part of GC+ is re-used (i.e. it's the same "sweep" that is already implemented for Azure) we should support this ootb?
> 
> There are active, production installations that we _know_ of that are running on Azure, with compliance requirements that are very hard to meet without GC+..

The sweep part is reused but we added more functionalities that weren't tesed on Azure

File: design/accepted/gc_plus/gc-plus-milestone2-execution-plan.md [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+    * Including GC changes, configuration, and behavior changes to fit GC
+3. Metrics and Logging additions
+4. Deployment to lakeFS Cloud
+5. Improve Backup & Restore
+
+### Plan
+
+* marks dependency
+
+1. Required changes by lakeFS:
```
Comment (itaiad200) [inline design/accepted/gc_plus/gc-plus-milestone2-execution-plan.md:34]: I think we're missing a non-trivial test issue here. I'm not sure how to test that lakeFS is safe while GC is running. Feels like we should plan for that here too.

File: design/accepted/gc_plus/gc-plus-milestone2-execution-plan.md [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+1. Support for non-S3 repositories
+   * Azure
+   * GCP
+2. Incorporation of committed & uncommitted GC into a single job
+    * Including GC changes, configuration, and behavior changes to fit GC
+3. Metrics and Logging additions
+4. Deployment to lakeFS Cloud
+5. Improve Backup & Restore
+
+### Plan
```
Comment (itaiad200) [inline design/accepted/gc_plus/gc-plus-milestone2-execution-plan.md:30]: Thanks for listing all the dependencies and the tasks themselves! It really made the difference for me and helped me understand what we're facing.

**Review by itaiad200 — CHANGES_REQUESTED** [review]
Thanks! It feels like a very-stressed milestone. I think we should settle for the "safe-write" lakeFS goal and leave the rest for the next milestones. The timeline itself is too short for all three, I think...

File: design/accepted/gc_plus/gc-plus-milestone2-execution-plan.md [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+
+### Goals
+1. Removing the limitation of a read-only lakeFS during the GC+ job run
+
+### Non-Goals
+1. Performance improvements:
+   * [Optimaized listing on old repository structure](https://github.com/treeverse/lakeFS/issues/4620)
+   * [Efficient listing on committed entries](https://github.com/treeverse/lakeFS/issues/4600)
+   * Benchmarks - verify uncommitted GC [performance requirements](https://github.com/treeverse/lakeFS/blob/e316cafe7717bb3203e4018837a41415aa61f74b/design/accepted/gc_plus/uncommitted-gc.md?plain=1#L185) are kept
+2. [Implement optimized run flow](https://github.com/treeverse/lakeFS/issues/4489)
```
Comment (N-o-Z) [inline design/accepted/gc_plus/gc-plus-milestone2-execution-plan.md:23]: Are we implementing the incremental run? I thought we decided to do it in the next milestone

File: design/accepted/gc_plus/gc-plus-milestone2-execution-plan.md [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+3. Support for non-S3 repositories
+   * Azure
+   * GCP
+4. Incorporation of committed & uncommitted GC into a single job
```
Comment (N-o-Z) [inline design/accepted/gc_plus/gc-plus-milestone2-execution-plan.md:27]: Same here - I thought we were going to push it towards the end of development

File: design/accepted/gc_plus/gc-plus-milestone2-execution-plan.md [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+    * [CopyObject API](https://github.com/treeverse/lakeFS/issues/4477)
+        * [S3 Gateway CopyObject](https://github.com/treeverse/lakeFS/issues/4478)
+        * [lakeFSFS renameObject method](https://github.com/treeverse/lakeFS/issues/4479)
+    * [Track copied objects in ref-store](https://github.com/treeverse/lakeFS/issues/4562)
+ 
+2. [Integration tests](https://github.com/treeverse/lakeFS/issues/4830) to verify lakeFS is safe while GC is running
+
+By the end of this milestone, we will release a new beta version that includes all the additions.
+
+**Due date: 15/01/2023**
```
Comment (N-o-Z) [inline design/accepted/gc_plus/gc-plus-milestone2-execution-plan.md:50]: IMO, this due date as a bit ambitious. We still left a lot of work inside this milestone

Comment (itaiad200) [reply design/accepted/gc_plus/gc-plus-milestone2-execution-plan.md:50]: It isn't anymore :)

**Review by itaiad200 — APPROVED** [review]
LGTM

**Review by N-o-Z — APPROVED** [review]
👍🙏

**Pull request merged**